### ISSUE TITLE
Function to run a top-level future on the thread-pool

### ISSUE THREAD

**Issue by Nemo157** [withoutboats/juliex#17]
I'm not sure how much sense this makes without a way to import the main thread into working for the thread pool (maybe that's possible?), but for consistency it would be nice to be able to use `juliex` from the top-level down, instead of the current example of using `futures::executor::block_on` to run the top-level future.

I tried using a basic utility function for this which just blocks the main thread waiting for the future to finish:

```rust
pub fn run<F>(future: F) -> F::Output where F: Future + Send + 'static, F::Output: Send + 'static {
    let (tx, rx) = std::sync::mpsc::channel();
    juliex::spawn(async move {
        let result = await!(future);
        tx.send(result).unwrap();
    });
    rx.recv().unwrap()
}
```

**Comment by yoshuawuyts** [comment]
@Nemo157 that's not a terrible idea tbh; the [`runtime-raw::enter`](https://github.com/rustasync/runtime/blob/master/runtime-raw/src/lib.rs#L55-L71) function does something very similar. In theory we could move some things around so `juliex` implements a version of this, and `runtime-raw`'s traits expect one of these to be implemented.

I reckon that would open us up for some performance improvements too later down the line, as the channel logic currently isn't great.

**Comment by najamelan** [comment]
Maybe I haven't understood entirely how to use juliex, but I feel like there is a bigger problem here. The futures library has 2 convenient methods, at least on the LocalPool: 
- run -> block until all spawned futures have finished
- run_until -> run until the topmost future has finished, ignoring other spawned tasks

This gives a level of control that juliex doesn't have:
```rust
fn main()
{
	use
	{
		async_runtime :: { *                                          } ,
		std           :: { time::{ Duration, Instant }, thread::sleep } ,
	};

	let program = async
	{
		let start = Instant::now();


		for i in 0..8
		{
			juliex::spawn( async move
			{
				sleep( Duration::SECOND );
				println!( "Time elapsed at task {} end: {} second(s).", i, start.elapsed().as_secs() );

			}).expect( "spawn task" );
		}
	};

	// futures::executor::block_on( program );
	run( program );

	// The program won't stay alive by itself
	//
	// sleep( Duration::SECOND * 2 );
}
```

In the example above, it doesn't matter if I use the `run` method suggested above or `block_on`. I suppose currently the most reasonable way to do this is to `join` all those futures, but if they would spawn more futures, all the way down there have to be synchronization mechanisms like channels to make sure the program doesn't exit prematurely...

Just had a look at the docs, it seems that for `futures::executor::ThreadPool` the [`run`](https://rust-lang-nursery.github.io/futures-api-docs/0.3.0-alpha.16/futures/executor/struct.ThreadPool.html#method.run) method corresponds to the [`run_until`](https://rust-lang-nursery.github.io/futures-api-docs/0.3.0-alpha.16/futures/executor/struct.LocalPool.html#method.run_until) method of `LocalPool`. A tad bit inconsistent. I couldn't find a way to run the `futures::executor::ThreadPool` until all tasks are finished.

**Comment by najamelan** [comment]
This is the most concise I have found so far (given that none of these tasks spawns any further):
```rust
use
{
	std           :: { time::{ Duration, Instant }, thread::sleep } ,
	futures       :: { future::{ FutureExt, join_all }            } ,
};

let program = async
{
	let start = Instant::now();
	let mut tasks = Vec::new();

	for i in 0..8
	{
		let (fut, handle) = async move
		{
			sleep( Duration::SECOND );

			println!( "Time elapsed at task {} end: {} second(s).", i, start.elapsed().as_secs() );

		}.remote_handle();

		rt::spawn( fut ).expect( "spawn task" );
		tasks.push( handle );
	}

	join_all( tasks ).await;
};

futures::executor::block_on( program );
```

**Comment by najamelan** [comment]
@Nemo157 Trying to understand your original post, why do you want to run the future on a different thread if you block the current thread anyway? Is it about juliex API having a method like block_on so people don't have to import futures::executor::block_on?

**Comment by Nemo157** [comment]
> Is it about juliex API having a method like block_on so people don't have to import futures::executor::block_on?

Yes, it feels weird to run different futures on different executors for no reason.

> I suppose currently the most reasonable way to do this is to join all those futures, but if they would spawn more futures, all the way down there have to be synchronization mechanisms like channels to make sure the program doesn't exit prematurely...

Personally I feel this is appropriate, the application should be concerned with which futures are ones it needs to wait on and which it should cancel if the main ones all complete. This can even be done transparently by passing around `Spawn` wrappers that register the spawned futures into collections instead of just spawning directly onto the executor.

Luckily where I was using Juliex I didn't have to worry about all that as the appropriate semantics if the top-level future exited was to just cancel all the spawned futures anyway.

**Comment by Nemo157** [comment]
Thinking about the `Spawn` wrappers idea a little more I wonder if it would be better to have an API like

```rust
pub fn run<F, Fut>(create: F) -> Fut::Output
where
    Fut: Future + Send + 'static,
    Fut::Output: Send + 'static,
    F: FnOnce(&juliex::Handle) -> Fut,
```

so that the top-level future can take in a `Spawn` instance to pass around.
